File: download_datasets.m
% -------------------------------------------------------------------------
% Section: Data Acquisition
% Purpose: Download and prepare datasets for compression experiments.
%          Fetches Text, Binary, and DNA data from public sources.
% Input:   None
% Output:  Files saved in 'data/' directory
% -------------------------------------------------------------------------

function download_datasets()
    if ~exist('data', 'dir')
        mkdir('data');
    end
    
    fprintf('=== Downloading Datasets ===\n');
    
    %% 1. Text Data: Alice in Wonderland (Project Gutenberg)
    url_text = 'https://www.gutenberg.org/files/11/11-0.txt';
    file_text = 'data/alice.txt';
    
    if ~exist(file_text, 'file')
        fprintf('Downloading Text Data (Alice in Wonderland)...\n');
        try
            websave(file_text, url_text);
            fprintf('  [OK] Saved to %s\n', file_text);
        catch ME
            fprintf('  [FAIL] Could not download text data: %s\n', ME.message);
            % Fallback: Create dummy text file
            fid = fopen(file_text, 'w');
            fprintf(fid, 'This is a fallback text file because the download failed.\n');
            for i = 1:100
                fprintf(fid, 'Repeating line %d for some volume.\n', i);
            end
            fclose(fid);
            fprintf('  [WARN] Created fallback text file.\n');
        end
    else
        fprintf('  [SKIP] Text data already exists.\n');
    end
    
    %% 2. Binary Data: System Binary (ls command)
    % Using a system binary ensures we have real executable code structure
    file_bin = 'data/binary_test.bin';
    source_bin = '/bin/ls'; % Standard on macOS/Linux
    
    if ~exist(file_bin, 'file')
        fprintf('Preparing Binary Data (Copying /bin/ls)...\n');
        if exist(source_bin, 'file')
            copyfile(source_bin, file_bin);
            fprintf('  [OK] Copied %s to %s\n', source_bin, file_bin);
        else
            % Fallback: Generate random binary data
            fprintf('  [WARN] /bin/ls not found. Generating random binary data.\n');
            rng(42);
            data = uint8(randi([0, 255], 1, 50000)); % 50KB
            fid = fopen(file_bin, 'w');
            fwrite(fid, data);
            fclose(fid);
            fprintf('  [OK] Generated random binary file.\n');
        end
    else
        fprintf('  [SKIP] Binary data already exists.\n');
    end
    
    %% 3. DNA Data: Synthetic Sequence
    % Downloading specific genome segments can be flaky due to URL changes.
    % We will generate a realistic synthetic DNA sequence.
    file_dna = 'data/dna.txt';
    
    if ~exist(file_dna, 'file')
        fprintf('Generating DNA Data (Synthetic)...\n');
        rng(123); % Reproducible
        bases = ['A', 'C', 'G', 'T'];
        len = 50000; % 50KB
        
        % Generate with some structure (e.g., GC-rich regions)
        dna_seq = char(zeros(1, len));
        for i = 1:len
            if mod(i, 100) < 50
                % GC-rich region
                weights = [0.1, 0.4, 0.4, 0.1];
            else
                % AT-rich region
                weights = [0.4, 0.1, 0.1, 0.4];
            end
            
            % Sample base
            r = rand;
            if r < weights(1)
                dna_seq(i) = bases(1);
            elseif r < weights(1) + weights(2)
                dna_seq(i) = bases(2);
            elseif r < weights(1) + weights(2) + weights(3)
                dna_seq(i) = bases(3);
            else
                dna_seq(i) = bases(4);
            end
        end
        
        fid = fopen(file_dna, 'w');
        fprintf(fid, '%s', dna_seq);
        fclose(fid);
        fprintf('  [OK] Generated synthetic DNA file (%d bases).\n', len);
    else
        fprintf('  [SKIP] DNA data already exists.\n');
    end
    
    fprintf('=== Data Acquisition Complete ===\n\n');
    
    % List files
    dir('data');
end

File: download_datasets.py
import os
import urllib.request
import tarfile
import shutil

def download_canterbury_corpus():
    """
    Downloads and extracts the Canterbury Corpus.
    """
    data_dir = "data"
    target_dir = os.path.join(data_dir, "canterbury")
    url = "http://corpus.canterbury.ac.nz/resources/cantrbry.tar.gz"
    tar_path = os.path.join(data_dir, "cantrbry.tar.gz")

    # Create directories
    if not os.path.exists(target_dir):
        os.makedirs(target_dir)
        print(f"Created directory: {target_dir}")

    # Download
    if not os.path.exists(tar_path):
        print(f"Downloading Canterbury Corpus from {url}...")
        try:
            # Add user agent to avoid 403 forbidden on some servers
            req = urllib.request.Request(
                url, 
                data=None, 
                headers={
                    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36'
                }
            )
            with urllib.request.urlopen(req) as response, open(tar_path, 'wb') as out_file:
                shutil.copyfileobj(response, out_file)
            print("Download complete.")
        except Exception as e:
            print(f"Error downloading file: {e}")
            return

    # Extract
    print("Extracting files...")
    try:
        with tarfile.open(tar_path, "r:gz") as tar:
            tar.extractall(path=target_dir)
        print(f"Extracted to {target_dir}")
    except Exception as e:
        print(f"Error extracting file: {e}")
        return

    # Cleanup tar file
    if os.path.exists(tar_path):
        os.remove(tar_path)
        print("Cleaned up temporary archive.")

    print("Canterbury Corpus ready.")

import random

def create_binary_test_file():
    """
    Creates a binary test file.
    Tries to copy /bin/ls (standard on macOS/Linux) for realistic executable structure.
    Falls back to random data if /bin/ls is not found.
    """
    data_dir = "data"
    file_bin = os.path.join(data_dir, "binary_test.bin")
    source_bin = "/bin/ls"

    if not os.path.exists(data_dir):
        os.makedirs(data_dir)

    if os.path.exists(file_bin):
        print("Binary test file already exists.")
        return

    print("Preparing Binary Data...")
    if os.path.exists(source_bin):
        try:
            shutil.copy(source_bin, file_bin)
            print(f"  [OK] Copied {source_bin} to {file_bin}")
        except Exception as e:
            print(f"  [WARN] Could not copy {source_bin}: {e}")
            _generate_random_binary(file_bin)
    else:
        print(f"  [WARN] {source_bin} not found.")
        _generate_random_binary(file_bin)

def _generate_random_binary(filepath):
    """Helper to generate random binary data."""
    print("  Generating random binary data...")
    try:
        size = 50000 # 50KB
        with open(filepath, "wb") as f:
            f.write(os.urandom(size))
        print("  [OK] Generated random binary file.")
    except Exception as e:
        print(f"  [FAIL] Could not generate random binary file: {e}")

if __name__ == "__main__":
    download_canterbury_corpus()
    create_binary_test_file()
